### ISSUE TITLE
BindingResolutionException on Connector if request is sent async

### ISSUE THREAD

**Issue by ahoiroman** [Sammyjo20/Saloon#320]
Hi there,

I got a connector, which resolves it's url this way:

```
    public function resolveBaseUrl(): string
    {
        $url = config('services.cloudservice.url');

        return "https://$url";
    }
```

If I use the connector using `send($request)` everything works like a charm. 

But if I do this: instead

```
return $this->cloudConnector->sendAsync($request)
```

I get this error:

> Illuminate\Contracts\Container\BindingResolutionException: Target class [config] does not exist

The trace points to:

```
->resolveBaseUrl(): string …
      › {
      ›     $url = config('services.cloudservice.url');
      ›

... 

/vendor/saloonphp/saloon/src/Http/PendingRequest.php:93 {▶
      Saloon\Http\PendingRequest->__construct(Connector $connector, Request $request, MockClient $mockClient = null) …
      › $this->method = $request->getMethod();
      › $this->url = URLHelper::join($this->connector->resolveBaseUrl(), $this->request->resolveEndpoint());
      › $this->authenticator = $request->getAuthenticator() ?? $connector->getAuthenticator();
    } 
```

**Comment by Sammyjo20** [comment]
Hey @ahoiroman 

Apologies for the delay in looking at this issue - this is certainly an odd one. Are you using Saloon in a Laravel application or a Laravel package? I've seen this issue before when the application is running in a package and Laravel's facades haven't been loaded yet. 

Any more information like Saloon version and any other saloonphp libraries installed would be useful

**Comment by ahoiroman** [comment]
Hi, not a problem, I am very thankful for your work.

I am using it within a Laravel app with PestPHP tests. The API calls are being called by a queued job.

**Comment by ahoiroman** [comment]
```
 composer show | grep saloon
saloonphp/laravel-plugin              v3.0.0    Laravel plugin for Saloon
saloonphp/saloon                      v3.0.2    Build beautiful API integra...
```

**Comment by Sammyjo20** [comment]
I wonder if this has to do with the promise wrapping that is happening inside of the `sendAsync` method. I'll have a look now.

**Comment by Sammyjo20** [comment]
I've been testing and I can't seem to replicate it - where is your 'cloudConnector' coming from inside of your job, does it come from the Laravel service container? If you are able to share some of your job class that would be great thank you

**Comment by ahoiroman** [comment]
Sure:

This is a part of my model class:

```
public function hcloud(): ?HcloudService
{
    if (!$this->hcloudToken) {
        return null;
    }

    $service = new HcloudService(hcloudToken: $this->hcloudToken);

    if ($this->hcloudSshKey_id) {
        $service->setHcloudSshKeyId(hcloudSshKeyId: $this->hcloudSshKey_id);
    }

    if ($this->hcloudServer_id) {
        $service->setHcloudServerId(hcloudServerId: $this->hcloudServer_id);
    }

    return $service;
}
```

And this is an example coming from my job:

```
$this->server
    ->hcloud()
    ->getSshKey()
    ->then(function () {
        $this->server
            ->hcloud()
            ->deleteSshKey()
            ->then(function () {
               //...
            });
    });

```


And this is coming from the service class:


```
public function getSshKey(int $hcloudSshKeyId = null): PromiseInterface|Response
{
    $hcloudSshKeyId = $hcloudSshKeyId ?? $this->hcloudSshKeyId;

    throw_if(condition: !$hcloudSshKeyId, exception: new InvalidArgumentException("No SSH ID provided"));

    $request = new GetSshKeyRequest($hcloudSshKeyId);

    return $this->hcloudConnector->sendAsync($request);
}

```

**Comment by Sammyjo20** [comment]
Thank you for sharing - apologies for asking you for lots of code, but are you able to share with me how `$this-. hcloudConnector` is being set?

**Comment by ahoiroman** [comment]
You don't have to apologize. It's me who is looking for help and I am thankful that you have a look at my code:

```
    public function __construct($hcloudToken)
    {
        $this->hcloudConnector = new HCloudConnector($hcloudToken);
    }
```

**Comment by juse-less** [comment]
@ahoiroman Hi!

Some questions:
1. Which version of Laravel are you using?
1. Is it possible to share the whole stack trace?
     You can omit parts like username, app name/path, if you want.
1. Does it fail during tests and/or when running the app?
     If in tests:
     1. Which version of PestPHP are you using?
     1. Can you share your test case class, or at least the imports and extend of it?

I've seen things like these happen before, both with helper functions, but also classes.
It's usually related to Laravel (and not something like Saloon), like extending the wrong base test case.
Or like Sam said, code trying to do things before dependent code has booted.

**Comment by Sammyjo20** [comment]
Hey @ahoiroman are you still having this issue?

**Comment by erikgaal** [comment]
Hey! Was running into this issue too when running Pest in parallel mode.

tl;dr I fixed it with the following snipped in a teardown of test cases:

```
public function tearDownSaloonTrait(): void
{
    Config::setSenderResolver(null);
    Config::clearGlobalMiddleware();
}
```

Why this works, or rather why it failed in the first place? The Saloon Laravel plugin uses the `Saloon\Config` to set global middleware and a sender resolver. These are `static` properties and will be shared in a parallel testing session between testcases. When you mix Laravel application tests with pure Unit tests, these config settings will persist. So in a Unit test, where you don't have the application booted, those middleware and sender resolver are still loaded and will try to access the application core.

In my opinion, Saloon should probably reset this `Config` class after each testcase, or give the tools to do this easily.

Cheers!

**Comment by Sammyjo20** [comment]
Ooh that's interesting @erikgaal thank you for sharing that. What version of Saloon and Laravel are you using? We're using L10 at work and Saloon v2 and not getting issues but it could be an issue with v3's service provider

**Comment by erikgaal** [comment]
```
▲ (erik.gaal) composer show | grep -E "saloon|laravel/framework|pestphp/pest"
laravel/framework                                v10.32.1            The Laravel Framework.
pestphp/pest                                     v2.23.2             The elegant PHP Testing Framework.
pestphp/pest-plugin                              v2.1.1              The Pest plugin manager
pestphp/pest-plugin-arch                         v2.4.0              The Arch plugin for Pest PHP.
pestphp/pest-plugin-laravel                      v2.2.0              The Pest Laravel Plugin
saloonphp/laravel-plugin                         v3.1.0              Laravel plugin for Saloon
saloonphp/saloon                                 v3.0.5              Build beautiful API integrations and SDKs with Saloon

**Comment by Sammyjo20** [comment]
Hey @erikgaal I've attempted to replicate this issue. I've created a brand new Laravel application with 60 individual tests (which are exact copies of each other) and I'm sending a Saloon request inside of each of them. It seems to work fine without the teardown in parallel and regular single-threaded tests.

<img width="1076" alt="Screenshot 2023-12-02 at 22 00 39" src="https://github.com/saloonphp/saloon/assets/29132017/e2833050-f0fc-432f-89bc-04c1cff49f6d">

This is the test which is run 60 times

```php
Saloon::fake([
   MockResponse::fixture('user'),
]);

$connector = new TestConnector;
$request = new UserRequest;

$response = $connector->send($request);

expect($response->status())->toEqual(200);
```

Let me know if I can refine my tests!

**Comment by Sammyjo20** [comment]
Hey @ahoiroman @erikgaal I'm going to close this issue to due inactivity and I don't believe this is an issue with Saloon. I'm happy to discuss this further if you want, but I'm going to close this for now.

**Comment by erikgaal** [comment]
Sorry, missed this. Did you mix Laravel's `TestCase` (that have `CreatesApplication`) and PHPUnit's base TestCase?

You need to mix tests with the Laravel application booting and tests that don't.

**Comment by Sammyjo20** [comment]
I will take a look, I'm pretty sure it was tests that booted Laravel, but I didn't mix tests with both, thanks for the info.

**Comment by Sammyjo20** [comment]
I'm going to close this one for now as it's really difficult replicate. Thanks all for helping out so far, hopefully in the future we can figure out the issue.
